### PR TITLE
ENH delta method and nonlinear prediction and test

### PR DESCRIPTION
**Pull request by josef-pkt**
- clean up and enhance `NonlinearDeltaCov`, move from miscmodels.count to stats
- add `LikelihoodModelResults._wald_test_nl`  as unit tested, experimental function, I'm not sure about API yet

see #2155 for delta method
??? issue for adding equivalent to `predict_nl` and `test_nl` in Stata
see also #2291 which has this PR as example or use case

`test_deltacov_margeff` shows how to redo a case of DiscreteMargins
### Todos, questions
- no unit tests for `NonlinearDeltaCov.wald_test`, coverage ?  - test comparing to linear and margins
- use_t: incompletely implemented, missing arguments and missing in wald_test
- add t_test to `NonlinearDeltaCov` as vectorized wald test
- caching: `NonlinearDeltaCov` does not cache results. My initial plan was that this is just a helper functions/class and the results are stored in the caller.
- add testing linear constraints on top of nonlinear transformation: wald_test and ttest in summary could test a Null hypothesis for linear restriction of the nonlinear prediction or transformation. Essentially redundant because it can be encoded in the nonlinear function, but might be convenient for some use cases. nothing for this is implemented.
- wrappers for specific use cases: `_wald_test_nl` is essentially a generic internal method. It's not clear to me yet how to make the API for user friendly versions of `predict_nl` and `test_nl`, especially we need the `at` and `atexog` (discrete margins) support.
- can we keep joint (wald_test) and vectorized (t_test) usages in the same method or would it be better to split? see

and
- compare with and unit test against Stata, check use cases

Naming: deriv, grad, or jac(obian) ?

### CODE REVIEW

File: statsmodels/stats/tests/test_deltacov.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        # test Results.method is same as calling function/class
+        res = self.res
+        x = res.model.exog
+
+        def fun(params):
+            return np.dot(x, params)**2
+
+        nl = NonlinearDeltaCov(fun, res.params, res.cov_params())
+        nlm = res._wald_test_nl(fun)
+        # margeff excludes constant, last parameter in this case
```
Comment (josef-pkt) [inline statsmodels/stats/tests/test_deltacov.py]: Copy paste noise

File: statsmodels/stats/_delta_method.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    def predicted(self):
+        # rename: misnomer, this is the MLE of the fun
+        predicted = self.fun(self.params)
+
+        # TODO: why do I need to squeeze in poisson example
+        if predicted.ndim > 1:
+            predicted = predicted.squeeze()
+        return predicted
+
+    def wald_test(self, value):
```
Comment (josef-pkt) [inline statsmodels/stats/_delta_method.py]: Add use_f

**Comment by josef-pkt** [comment]
last commit shouldn't [skip ci]  need tests for keyword argument renaming

File: statsmodels/stats/_delta_method.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+from __future__ import print_function
+import numpy as np
+from scipy import stats
+
+class NonlinearDeltaCov(object):
+    '''Asymptotic covariance by Deltamethod
+
+    the function is designed for 2d array, with rows equal to
+    the number of equations and columns equal to the number
```
Comment (josef-pkt) [inline statsmodels/stats/_delta_method.py]: it's a class now not a function

Can parts be moved or shared with ContrastResults, like the model Results.t_test results?

**Comment by josef-pkt** [comment]
separate internal helper function from user facing API decision. Merge basic functionality, get applications and enhancements later. (I need PRs where I don't need a long time to make up my mind.)

test coverage looks ok, so throw it in

**Comment by josef-pkt** [comment]
rebased, no merge conflicts

**Comment by josef-pkt** [comment]
rebase didn't trigger unit test runs, 
it's because the last commit includes [skip ci]

**Comment by josef-pkt** [comment]
according to last comments above, the basics should be ready to merge

Needs a rebase and code commit to trigger CI testing.

related: 

`eform`, univariate monotonic transformation of individual params is in master but not used and not advertised
https://github.com/statsmodels/statsmodels/issues/2235#issuecomment-623154082

What's the status on supporting pre-specified transformations, like like links or inverse_links/transforms?

another possible application:
- predict_prob in count models and related. Those are nonlinear functions of parameters and don't have `get_prediction` yet. #7601

**Comment by josef-pkt** [comment]
merged in PR #7758
